src/array.h

```diff
@@ -0,0 +1,86 @@
+#ifndef __ARRAY_H_
+#define __ARRAY_H_
+
+#include "zmalloc.h"
+
+/* array.h - A generic dynamic array implementation.
+ * Similar to SDS, the array's metadata and data are stored contiguously.
+ * So the actual array layout is this one:
+ * +-------------+-------------+---------+-----+-----------------------\
+ * | len | alloc | free method | element | ... | element | Free space  \
+ * +-------------+-------------+---------+-----+-----------------------\
+ *                             |
+ *                             `-> Pointer returned to the user.
+ */
+
+/* Metadata struct of array */
+typedef struct array {
+    unsigned long len;       /* Current length of the array */
+    unsigned long alloc;     /* Allocated size of the array */
+    void (*free)(void *ptr); /* Free method for each element */
+    char data[];             /* Pointer returned to the user */
+} array;
+
+/* Helper Macros and Internal Functions */
+#define ARRAY_META_SIZE sizeof(array)
+
+/* Convert the data pointer to the array metadata pointer */
+#define ARRAY_HDR(v) (((array *)(v)) - 1)
+
+/* It is an internal function that expands the array before pushing each element
+ * and return the data pointer */
+static void *arrayGrowIfNeed(array *a, unsigned long item_size) {
+    if (a->len < a->alloc) {
+        a->len += 1;
+        return a->data;
+    }
+    a->alloc = a->alloc ? a->alloc * 2 : 8;
+    a = zrealloc(a, a->alloc * item_size + ARRAY_META_SIZE);
+    a->len += 1;
+    return a->data;
+}
+
+/* Exported APIs
+ * arrayNew, arrayPush, and other APIs are implemented via macros,
+ * which makes the code harder to read, primarily for two reasons:
+ *   1. To obtain the element size at compile time, avoiding the
+ *      need to store element size in the array's metadata;
+ *   2. When used, it behaves more like a static array. For example,
+ *      we can access individual elements via dynamic_array[index].
+ */
+#define arrayNew(v, s, f)                                                          \
+    do {                                                                           \
+        array *__array_meta_ptr__ = zmalloc(ARRAY_META_SIZE + sizeof(*(v)) * (s)); \
+        __array_meta_ptr__->len = 0;                                               \
+        __array_meta_ptr__->alloc = (s);                                           \
+        __array_meta_ptr__->free = (f);                                            \
+        v = (void *)(__array_meta_ptr__ + 1);                                      \
+    } while (0)
```

Comment:
Macro code like this is difficult to read and not type safe.

If we want to write this in the header file, it can be done as `static inline` functions. They are just like macros but easier to read. But there is not really a reason. We can put the implementation in array.c. We compile with link-time-optimization (-flto) by default so they can still be inlined in other C files.